### PR TITLE
Add string like filtering capabilities to filter widget

### LINKED ISSUE #40770: [49.1] impossible to filter on IDs identified as Identity key (PK)
### Describe the bug

impossible to filter on IDs identified as Identity key (PK)
image

### To Reproduce

1. Go to a collection with a field set as Primary key by Metabase discovery
2. Click on filters
3. Scroll down to ID

### Expected behavior

Filtering on IDs should be possible

### Logs

_No response_

### Information about your Metabase installation

```JSON
image: metabase/metabase:v0.49.1
MB_DB_TYPE: mysql
```

### Severity

Low

### Additional context

_No response_

### LINKED ISSUE #42010: Unable to filter on MongoDB ObjectId type fields if configured as Entity Key
### Describe the bug

MongoDB ObjectId type fields set as Entity Key do nothing when selected for filtering in the notebook editor.

Similar to https://github.com/metabase/metabase/issues/38288 but happening in GUI questions as well.

Clicking here:
image

Does nothing:
image

### To Reproduce

1. Add [MongoDB Sample Database](https://hub.docker.com/layers/metabase/qa-databases/mongo-sample-5/images/sha256-0d0d0199277d88a4f10de5f234e5101db62378381736e476979da4a8a69eb54b?context=explore) as data source
2. New > Question > Raw Data > `MongoDB Sample Database from qa-databases` > Accounts
3. Click Add filters or narrow your answer
4. Click the first ID in the list (see note below)
5. Nothing happens (this is the bug)

If ObjectId `_id` field set as `No semantic type` or `Category` instead of `Entity Key`, the filter works.

NOTE: There is another ID in the list, but it is a different data type. The one that does not work is `_id` which is `type/MongoBSONID`.
image

### Expected behavior

- Should be able to filter on Entity Keys with MongoDB like on any other supported database Entity Key

### Logs

no errors on frontend or backend logs

### Information about your Metabase installation

```JSON
- Broken in Metabase v1.49.7 and v1.49.6
- Worked in 1.48.10
```

### Severity

P1: Common functionality not working, regression

### Additional context

Related to:
- https://github.com/metabase/metabase/issues/38288 (refers to native queries only)

May be related to:
- https://github.com/metabase/metabase/issues/38469
- https://github.com/metabase/metabase/issues/38311

### PR DESCRIPTION
**Pull request by lbrdnk**
Closes [#42010](https://github.com/metabase/metabase/issues/42010)
Closes [#40770](https://github.com/metabase/metabase/issues/40770)

The PR introduces E2E reproductions for both issues.

This PR enables string filtering widget for `string_like` columns.

### CODE REVIEW

File: frontend/src/metabase/querying/components/FilterPicker/FilterPickerBody/FilterPickerBody.tsx [hunk 1]
```diff
@@ -60,7 +60,7 @@ function getFilterWidget(column: Lib.ColumnMetadata) {
   if (Lib.isNumeric(column)) {
     return NumberFilterPicker;
   }
-  if (Lib.isString(column)) {
+  if (Lib.isString(column) || Lib.isStringLike(column)) {
```
Comment (ranquild) [inline frontend/src/metabase/querying/components/FilterPicker/FilterPickerBody/FilterPickerBody.tsx:63]: You also need to update
- `stringFilterParts` https://github.com/metabase/metabase/blob/4b71e9aa645117a944af409882bda33f64225bf9/frontend/src/metabase-lib/filter.ts#L139
- `ColumnFilterSection` https://github.com/metabase/metabase/blob/2aa3cec610588238132afbc789da0845c7c2ec5d/frontend/src/metabase/querying/components/FilterModal/ColumnFilterSection/ColumnFilterSection.tsx#L64

Comment (ranquild) [reply frontend/src/metabase/querying/components/FilterPicker/FilterPickerBody/FilterPickerBody.tsx:63]: Maybe we should export a function `isStringOrStringLike` from `column_types.ts` if it becomes a common operation.

**Comment by nemanjaglumac** [comment]
We need to check for three scenarios in E2E reproduction to be reasonably safe against a regression.

### Simple Mode (aka chill mode)
Scenario 1
1. Click on an id directly in the table
2. Make sure it's applied
3. Click on the purple filter value "pill" saying "ID is 66...."
4. Make sure that the filter picker is not showing a custom expression, but rather a string filter
Example of how it shouldn't work 
image

Scenario 2
1. Click on the Filter
2. Make sure we can enter ID in the filter modal
3. Add filter and make sure it's working
Example of what's currently going on in master
image

### Notebook mode
1. Click on the Filter
2. Select the first ID
3. Make sure the popover shows up with the ability to enter a string filter

File: e2e/test/scenarios/question/reproductions/42065-mongo-filter-by-id.cy.spec.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        openTable({
+          database: WRITABLE_DB_ID,
+          table: tableId,
+          mode: "notebook",
+        });
+      });
+    });
+
+    it("should be possible to filter by Mongo _id column (metabase#42010)", () => {
+      cy.get("button").contains("Visualize").click();
```
Comment (nemanjaglumac) [inline e2e/test/scenarios/question/reproductions/42065-mongo-filter-by-id.cy.spec.js:23]: If you omit the `mode: "notebook"` from the `openTable`, you can remove this step.

File: e2e/test/scenarios/question/reproductions/42065-mongo-filter-by-id.cy.spec.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+      restore("mongo-5");
+      cy.signInAsAdmin();
+
+      cy.request(`/api/database/${WRITABLE_DB_ID}/schema/`).then(({ body }) => {
+        const tableId = body.find(table => table.name === "orders").id;
+        openTable({
+          database: WRITABLE_DB_ID,
+          table: tableId,
+          mode: "notebook",
+        });
```
Comment (nemanjaglumac) [inline e2e/test/scenarios/question/reproductions/42065-mongo-filter-by-id.cy.spec.js:18]:
```suggestion
        openTable({
          database: WRITABLE_DB_ID,
          table: tableId,
          limit: 2,
        });
```

Comment (nemanjaglumac) [reply e2e/test/scenarios/question/reproductions/42065-mongo-filter-by-id.cy.spec.js:18]: We're adding a limit for performance improvements in CI. Every bit helps.

File: e2e/test/scenarios/question/reproductions/42065-mongo-filter-by-id.cy.spec.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+          cy.log("Check notebook editor");
+          cy.findByRole("button", { name: "Filter" }).click();
+          // Following selects first item from filter column picker
+          cy.get("[data-element-id=list-item-title]")
+            .eq(0)
+            .should("have.text", "ID");
+          cy.get("[data-element-id=list-item-title]").eq(0).click();
+          cy.findByPlaceholderText("Search by ID").type(id);
+          cy.get("button").contains("Add filter").click();
+          cy.get("button").contains("Visualize").click();
```
Comment (nemanjaglumac) [inline e2e/test/scenarios/question/reproductions/42065-mongo-filter-by-id.cy.spec.js:42]: There is a helper for this that waits for the dataset under the hood. Please use it instead.
```suggestion
          visualize();
```

File: e2e/test/scenarios/question/reproductions/42065-mongo-filter-by-id.cy.spec.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+          cy.go("back");
+
+          // Check the notebook editor
+          cy.log("Check notebook editor");
+          cy.findByRole("button", { name: "Filter" }).click();
+          // Following selects first item from filter column picker
+          cy.get("[data-element-id=list-item-title]")
+            .eq(0)
+            .should("have.text", "ID");
+          cy.get("[data-element-id=list-item-title]").eq(0).click();
```
Comment (nemanjaglumac) [inline e2e/test/scenarios/question/reproductions/42065-mongo-filter-by-id.cy.spec.js:39]: You can (and should) chain the `.click()` directly on the previous selector. No need to "get it" again.

But more than that, we should scope this to the popover that appears.
```suggestion
          popover()
            .findAllByRole("option").first()
            .should("have.text", "ID")
            .click();
```

File: e2e/test/scenarios/question/reproductions/42065-mongo-filter-by-id.cy.spec.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+          cy.findByPlaceholderText("Search by ID").type(id);
+          cy.get("button").contains("Add filter").click();
```
Comment (nemanjaglumac) [inline e2e/test/scenarios/question/reproductions/42065-mongo-filter-by-id.cy.spec.js:41]:
```suggestion
          cy.button("Add filter").click();
```

File: e2e/test/scenarios/question/reproductions/42065-mongo-filter-by-id.cy.spec.js [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+          cy.get("button").contains("Visualize").click();
+          cy.get("button").contains("Showing 1 row");
```
Comment (nemanjaglumac) [inline e2e/test/scenarios/question/reproductions/42065-mongo-filter-by-id.cy.spec.js:43]:
```suggestion
          cy.findByTestId("question-row-count").should("have.text", "Showing 1 row");
```

File: e2e/test/scenarios/question/reproductions/42065-mongo-filter-by-id.cy.spec.js [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+          // Now remove the filter
+          cy.log("Remove current filter");
```
Comment (nemanjaglumac) [inline e2e/test/scenarios/question/reproductions/42065-mongo-filter-by-id.cy.spec.js:46]: `cy.log()` is good enough here. Feel free to remove the inline comment.

```suggestion
          cy.log("Remove current filter");
```

File: e2e/test/scenarios/question/reproductions/42065-mongo-filter-by-id.cy.spec.js [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+    it("should be possible to filter by Mongo _id column (metabase#42010)", () => {
+      cy.get("button").contains("Visualize").click();
+      cy.get("[data-testid=cell-data]")
+        .eq(10)
+        .then(elm => {
+          // Get the id of some row and go back to notebook editor. I suppose it differs per snapshot, hence we are
+          // unable to hardcode the value.
+          const id = elm.text();
+          cy.go("back");
```
Comment (nemanjaglumac) [inline e2e/test/scenarios/question/reproductions/42065-mongo-filter-by-id.cy.spec.js:30]: This conveys what we're doing more explicitly to anyone reading the code.
```suggestion
          openNotebook();
```

File: frontend/src/metabase/querying/components/FilterPicker/FilterPickerBody/FilterPickerBody.tsx (hunk 1 above)
Comment (nemanjaglumac) [inline frontend/src/metabase/querying/components/FilterPicker/FilterPickerBody/FilterPickerBody.tsx:63]: @lbrdnk Is there anything preventing us from using the new abstraction here as well?
```suggestion
  if (Lib.isStringOrStringLike(column) {
```

**Review by nemanjaglumac — APPROVED** [review]
Left just one question/comment.
Tested locally, and it works as advertised. Both issues are solved with this approach.

Thank you

**Pull request merged**